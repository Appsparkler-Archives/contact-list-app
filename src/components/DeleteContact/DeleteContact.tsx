import { TDeleteContactModalFC } from "types";
import { Delete as DeleteIcon } from "@mui/icons-material";
import * as React from "react";
import Button from "@mui/material/Button";
import Typography from "@mui/material/Typography";
import { AppBar, Box, Dialog, IconButton, Toolbar } from "@mui/material";
import { Close } from "@mui/icons-material";
import { FullScreenModalTransition } from "components/FullScreenModalTransition/FullScreenModalTransition";

export const DeleteContact: TDeleteContactModalFC = ({ contact, onDelete }) => {
  const [open, setOpen] = React.useState(false);

  const handleOpen = () => setOpen(true);

  const handleClose = () => setOpen(false);

  const handleDelete: React.MouseEventHandler<HTMLButtonElement> =
    React.useCallback(() => {
      handleClose();
      setTimeout(() => {
        onDelete(contact.id);
      }, 500);
    }, [contact, onDelete]);

  return (
    <Box>
      <IconButton aria-label="delete" color="warning" onClick={handleOpen}>
        <DeleteIcon />
      </IconButton>
      <Dialog
        open={open}
        onClose={handleClose}
        fullScreen
        TransitionComponent={FullScreenModalTransition}
      >
        <AppBar position="static" sx={{ backgroundColor: "warning.main" }}>
          <Toolbar>
            <IconButton
              edge="start"
              color="inherit"
              onClick={handleClose}
              aria-label="close"
            >
              <Close />
            </IconButton>
            <Typography sx={{ ml: 2, flex: 1 }} variant="h6" component="div">
              Delete Contact
            </Typography>
          </Toolbar>
        </AppBar>
        <Typography p={2} gutterBottom variant="body1" component="div">
          Are you sure you want to delete <strong>{contact.name}</strong> from
          your contact list?
        </Typography>
        <Box
          display="flex"
          flexDirection={"row"}
          justifyContent={"flex-end"}
          p={2}
          gap={2}
        >
          <Button
            size="large"
            variant="outlined"
            color="warning"
            onClick={handleClose}
          >
            Cancel
          </Button>
          <Button
            size="large"
            color="warning"
            variant="contained"
            onClick={handleDelete}
          >
            Yes, delete
          </Button>
        </Box>
      </Dialog>
    </Box>
  );
};
