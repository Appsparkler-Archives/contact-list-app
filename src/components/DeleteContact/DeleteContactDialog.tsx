import * as React from "react";
import Button from "@mui/material/Button";
import Typography from "@mui/material/Typography";
import { AppBar, Box, Dialog, IconButton, Toolbar } from "@mui/material";
import { Close } from "@mui/icons-material";
import { TDeleteContactDialogFC } from "types";

export const DeleteContactDialog: TDeleteContactDialogFC = ({
  contact,
  open,
  onClose,
  onDelete,
  onCancel,
}) => {
  const handleDelete: (
    idToDelete: string
  ) => React.MouseEventHandler<HTMLButtonElement> = React.useCallback(
    (idToDelete) => () => {
      onDelete(idToDelete);
    },
    [onDelete]
  );
  return (
    <Dialog open={open} onClose={onClose} fullScreen>
      <AppBar position="static">
        <Toolbar>
          <IconButton
            edge="start"
            color="inherit"
            onClick={onClose}
            aria-label="close"
          >
            <Close />
          </IconButton>
          <Typography sx={{ ml: 2, flex: 1 }} variant="h6" component="div">
            Delete Contact
          </Typography>
        </Toolbar>
      </AppBar>
      <Typography p={2} gutterBottom variant="body1" component="div">
        Are you sure you want to delete <strong>{contact.name}</strong> from
        your contact list?
      </Typography>
      <Box
        display="flex"
        flexDirection={"row"}
        justifyContent={"flex-end"}
        p={2}
        gap={2}
      >
        <Button
          size="large"
          variant="outlined"
          color="primary"
          onClick={onCancel}
        >
          Cancel
        </Button>
        <Button
          size="large"
          color="error"
          variant="contained"
          onClick={handleDelete(contact.id)}
        >
          Yes, delete
        </Button>
      </Box>
    </Dialog>
  );
};
