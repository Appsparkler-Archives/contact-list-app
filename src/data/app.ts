import { IContactFormData } from "types";

export type TAppFC = React.FC<{
  contacts: IContactFormData[];
}>;
