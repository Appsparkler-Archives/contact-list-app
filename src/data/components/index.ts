export * from "./standard-select-field";
