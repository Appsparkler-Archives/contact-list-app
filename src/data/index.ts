export * from "./contacts";
export * from "./components";
