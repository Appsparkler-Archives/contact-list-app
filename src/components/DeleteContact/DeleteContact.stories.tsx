import type { Meta, StoryObj } from "@storybook/react";
import { DeleteContact } from "./DeleteContact";
import { contactsData } from "data";

const meta = {
  title: "Components/Delete Contact Modal",
  component: DeleteContact,
  tags: ["components"],
} satisfies Meta<typeof DeleteContact>;

export default meta;

type Story = StoryObj<typeof meta>;

export const deleteContact: Story = {
  args: {
    contact: contactsData.variant1[0],
  },
};
