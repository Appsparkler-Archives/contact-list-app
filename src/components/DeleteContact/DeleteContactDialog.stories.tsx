import type { Meta, StoryObj } from "@storybook/react";
import { DeleteContactDialog } from "./DeleteContactDialog";
import { contactsData } from "data";

const meta = {
  title: "Components/Delete Contact Modal/Delete Contact Dialog",
  component: DeleteContactDialog,
  tags: ["components"],
} satisfies Meta<typeof DeleteContactDialog>;

export default meta;

type Story = StoryObj<typeof meta>;

export const deleteContactDialog: Story = {
  args: {
    contact: contactsData.variant1[0],
    open: true,
  },
};
