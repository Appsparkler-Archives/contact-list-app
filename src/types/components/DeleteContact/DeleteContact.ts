import React from "react";
import { IContactFormData } from "types/data";

export type TDeleteContactDialogProps = {
  open: boolean;
  contact: IContactFormData;
  onClose: () => void;
  onDelete: (contactIdToDelete: string) => void;
  onCancel: React.MouseEventHandler<HTMLButtonElement>;
};

export type TDeleteContactModalFC = React.FC<{
  contact: IContactFormData;
  onDelete: (contactIdToDelete: string) => void;
}>;
