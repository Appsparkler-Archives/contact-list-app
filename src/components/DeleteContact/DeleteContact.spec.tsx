import { act, fireEvent, render, screen } from "@testing-library/react";
import { DeleteContact } from "./DeleteContact";
import { IContactFormData } from "types";
import { contactsData } from "data";

const contact2Delete: IContactFormData = contactsData.variant1[0];
const handleDelete = jest.fn();
jest.useFakeTimers();

describe("DeleteContact Snapshots", () => {
  it("renders closed modal correctly", () => {
    const { container } = renderDeleteContactComponent();
    expect(container).toMatchSnapshot();
  });

  it("renders open modal correctly", () => {
    const { baseElement } = renderDeleteContactComponent();
    clickOnDeleteIconButton();
    expect(baseElement).toMatchSnapshot();
  });
});

describe("DeleteCOntact - functional tests", () => {
  it("calls onDelete when delete button is clicked", () => {
    renderDeleteContactComponent();
    clickOnDeleteIconButton();
    clickOnYesDeleteButton();
    act(() => {
      jest.runAllTimers();
    });
    expect(handleDelete).toHaveBeenCalled();
  });
});

function renderDeleteContactComponent(): { container: any; baseElement: any } {
  return render(
    <DeleteContact contact={contact2Delete} onDelete={handleDelete} />
  );
}

function clickOnDeleteIconButton() {
  const deleteButton = screen.getByLabelText(/delete/);
  fireEvent.click(deleteButton);
}

function clickOnYesDeleteButton() {
  const yesDeleteButton = screen.getByText("Yes, delete");
  fireEvent.click(yesDeleteButton);
}
