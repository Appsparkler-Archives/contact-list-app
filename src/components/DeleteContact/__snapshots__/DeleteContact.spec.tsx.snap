// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DeleteContact Snapshots renders closed modal correctly 1`] = `
<div>
  <div
    class="MuiBox-root css-0"
  >
    <button
      aria-label="delete"
      class="MuiButtonBase-root MuiIconButton-root MuiIconButton-colorWarning MuiIconButton-sizeMedium css-1moznky-MuiButtonBase-root-MuiIconButton-root"
      tabindex="0"
      type="button"
    >
      <svg
        aria-hidden="true"
        class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
        data-testid="DeleteIcon"
        focusable="false"
        viewBox="0 0 24 24"
      >
        <path
          d="M6 19c0 1.1.9 2 2 2h8c1.1 0 2-.9 2-2V7H6v12zM19 4h-3.5l-1-1h-5l-1 1H5v2h14V4z"
        />
      </svg>
      <span
        class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
      />
    </button>
  </div>
</div>
`;

exports[`DeleteContact Snapshots renders open modal correctly 1`] = `
<body
  style="padding-right: 1024px; overflow: hidden;"
>
  <div
    aria-hidden="true"
  >
    <div
      class="MuiBox-root css-0"
    >
      <button
        aria-label="delete"
        class="MuiButtonBase-root MuiIconButton-root MuiIconButton-colorWarning MuiIconButton-sizeMedium css-1moznky-MuiButtonBase-root-MuiIconButton-root"
        tabindex="0"
        type="button"
      >
        <svg
          aria-hidden="true"
          class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
          data-testid="DeleteIcon"
          focusable="false"
          viewBox="0 0 24 24"
        >
          <path
            d="M6 19c0 1.1.9 2 2 2h8c1.1 0 2-.9 2-2V7H6v12zM19 4h-3.5l-1-1h-5l-1 1H5v2h14V4z"
          />
        </svg>
        <span
          class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
        />
      </button>
    </div>
  </div>
  <div
    class="MuiDialog-root MuiModal-root css-zw3mfo-MuiModal-root-MuiDialog-root"
    role="presentation"
  >
    <div
      aria-hidden="true"
      class="MuiBackdrop-root MuiModal-backdrop css-yiavyu-MuiBackdrop-root-MuiDialog-backdrop"
      style="opacity: 1; webkit-transition: opacity 225ms cubic-bezier(0.4, 0, 0.2, 1) 0ms; transition: opacity 225ms cubic-bezier(0.4, 0, 0.2, 1) 0ms;"
    />
    <div
      data-testid="sentinelStart"
      tabindex="0"
    />
    <div
      class="MuiDialog-container MuiDialog-scrollPaper css-hz1bth-MuiDialog-container"
      role="presentation"
      style="webkit-transform: none; transform: none; webkit-transition: -webkit-transform 225ms cubic-bezier(0.0, 0, 0.2, 1) 0ms; transition: transform 225ms cubic-bezier(0.0, 0, 0.2, 1) 0ms;"
      tabindex="-1"
    >
      <div
        aria-labelledby=":r1:"
        class="MuiPaper-root MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation24 MuiDialog-paper MuiDialog-paperScrollPaper MuiDialog-paperWidthSm MuiDialog-paperFullScreen css-m9glnp-MuiPaper-root-MuiDialog-paper"
        role="dialog"
      >
        <header
          class="MuiPaper-root MuiPaper-elevation MuiPaper-elevation4 MuiAppBar-root MuiAppBar-colorPrimary MuiAppBar-positionStatic css-1efs0ro-MuiPaper-root-MuiAppBar-root"
        >
          <div
            class="MuiToolbar-root MuiToolbar-gutters MuiToolbar-regular css-hyum1k-MuiToolbar-root"
          >
            <button
              aria-label="close"
              class="MuiButtonBase-root MuiIconButton-root MuiIconButton-colorInherit MuiIconButton-edgeStart MuiIconButton-sizeMedium css-1d6w9lk-MuiButtonBase-root-MuiIconButton-root"
              tabindex="0"
              type="button"
            >
              <svg
                aria-hidden="true"
                class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
                data-testid="CloseIcon"
                focusable="false"
                viewBox="0 0 24 24"
              >
                <path
                  d="M19 6.41 17.59 5 12 10.59 6.41 5 5 6.41 10.59 12 5 17.59 6.41 19 12 13.41 17.59 19 19 17.59 13.41 12z"
                />
              </svg>
              <span
                class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
              />
            </button>
            <div
              class="MuiTypography-root MuiTypography-h6 css-6b3s0t-MuiTypography-root"
            >
              Delete Contact
            </div>
          </div>
        </header>
        <div
          class="MuiTypography-root MuiTypography-body1 MuiTypography-gutterBottom css-y5gjj9-MuiTypography-root"
        >
          Are you sure you want to delete 
          <strong>
            Ali Connors
          </strong>
           from your contact list?
        </div>
        <div
          class="MuiBox-root css-y041cw"
        >
          <button
            class="MuiButtonBase-root MuiButton-root MuiButton-outlined MuiButton-outlinedWarning MuiButton-sizeLarge MuiButton-outlinedSizeLarge MuiButton-root MuiButton-outlined MuiButton-outlinedWarning MuiButton-sizeLarge MuiButton-outlinedSizeLarge css-1fmotu4-MuiButtonBase-root-MuiButton-root"
            tabindex="0"
            type="button"
          >
            Cancel
            <span
              class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
            />
          </button>
          <button
            class="MuiButtonBase-root MuiButton-root MuiButton-contained MuiButton-containedWarning MuiButton-sizeLarge MuiButton-containedSizeLarge MuiButton-root MuiButton-contained MuiButton-containedWarning MuiButton-sizeLarge MuiButton-containedSizeLarge css-1osjee8-MuiButtonBase-root-MuiButton-root"
            tabindex="0"
            type="button"
          >
            Yes, delete
            <span
              class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
            />
          </button>
        </div>
      </div>
    </div>
    <div
      data-testid="sentinelEnd"
      tabindex="0"
    />
  </div>
</body>
`;
